Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,414 @@ extension MongoCollection {
      */
     @discardableResult
     public func bulkWrite(_ requests: [WriteModel], options: BulkWriteOptions? = nil) throws -> BulkWriteResult? {
-        throw MongoError.commandError(message: "Unimplemented command")
+        if requests.isEmpty {
+            throw MongoError.invalidArgument(message: "requests cannot be empty")
+        }
+
+        let opts = try BsonEncoder().encode(options)
+        let bulk = BulkWriteOperation(collection: self._collection, opts: opts?.data)
+
+        try requests.enumerated().forEach { (index, model) in
+            try model.addToBulkWrite(bulk: bulk, index: index)
+        }
+
+        return try bulk.execute()
     }
-}
 
-/// Options to use when performing a bulk write operation on a `MongoCollection`.
-public struct BulkWriteOptions: Encodable {
-    /// If `true` (the default), when an insert fails, return without performing the remaining writes.
-    /// If `false`, when a write fails, continue with the remaining writes, if any.
-    public var ordered: Bool = true
+    private struct DeleteModelOptions: Encodable {
+        public let collation: Document?
+    }
 
-    /// If `true`, allows the write to opt-out of document level validation.
-    public let bypassDocumentValidation: Bool?
-}
+    /// A model for a `deleteOne` operation within a bulk write.
+    public struct DeleteOneModel: WriteModel {
+        private let filter: Document
+        private let options: DeleteModelOptions
 
-/// A protocol indicating write types that can be batched together using `MongoCollection.bulkWrite`.
-public protocol WriteModel: Encodable {}
+        /**
+         * Create a `deleteOne` operation for a bulk write.
+         *
+         * - Parameters:
+         *   - filter: A `Document` representing the match criteria
+         *   - collation: Specifies a collation to use
+         */
+        public init(_ filter: Document, collation: Document? = nil) {
+            self.filter = filter
+            self.options = DeleteModelOptions(collation: collation)
+        }
 
-/// A model for an `insertOne` command.
-public struct InsertOneModel: WriteModel {
-    /// The `Document` to insert.
-    public let document: Document
-}
+        /// Adds the `deleteOne` operation to a bulk write
+        public func addToBulkWrite(bulk: BulkWriteOperation, index: Int) throws {
+            let opts = try BsonEncoder().encode(options)
+            var error = bson_error_t()
 
-/// A model for a `deleteOne` command.
-public struct DeleteOneModel: WriteModel {
-    /// A `Document` representing the match criteria.
-    public let filter: Document
+            guard mongoc_bulk_operation_remove_one_with_opts(bulk._bulk, filter.data, opts.data, &error) else {
```

Comment:
> you are breaking our linting rule about explicitly using self in these cases

Interesting, I didn't realize this was a lint violation. Sublime Text tends to inject a warning into the gutter when I have violations and this was clear. Happy to fix.

> I think this line, and many like it in the rest of the PR, should accommodate the fact that opts might be nil by doing opts?.data.

`opts` is initialized a few lines up with `let opts = try BsonEncoder().encode(options)`. `options` is always non-nil, as it's initialized to a DeleteModelOptions instance in `init`. When we encode, we're using the [`encode<T: Encodable>(_ value: T) throws -> Document`](https://github.com/mongodb/mongo-swift-driver/blob/b5f7e5783da71270fa80f5bad7c28159dab3e1e6/Sources/MongoSwift/BSON/BsonEncoder.swift#L28) variant instead of [`encode<T: Encodable>(_ value: T?) throws -> Document?`](https://github.com/mongodb/mongo-swift-driver/blob/b5f7e5783da71270fa80f5bad7c28159dab3e1e6/Sources/MongoSwift/BSON/BsonEncoder.swift#L64), so nil is never returned.

That said, I think `try BsonEncoder().encode(options)` is probably a violation of the lint rule you mentioned and should be changed to `self.options`.

Reply:
see: https://github.com/mongodb/mongo-swift-driver/pull/97/files/28a86dc11fccae56f2030606a148217285f9b8c6#diff-312fa0fd8cb91f4e116a176fe7b0133eR25

Reply:
Oh I see, yeah. Either way let's decide on whether to explicitly use `self` or not, I think it makes it harder to read without

Reply:
we don't have a rule for that atm, but there is an [explicit self](https://github.com/realm/SwiftLint/blob/master/Rules.md#explicit-self) rule we can add. either it's new or I missed it on my last pass through the list, would've added it

Reply:
Will fix the implicit self here, but we'll handle it across the board in [SWIFT-217](https://jira.mongodb.org/browse/SWIFT-217).